File: scripts/01_data_cleaning.R
# ============================================================
# IT3011 Group Assignment - Script 01: Data Cleaning
# Learning Outcomes: LO1, LO4
# Analytical Statement: "students who practice self-assessment develop stronger learning skills"
# ============================================================

# Load libraries
library(tidyverse)

# ---- 1. Load Raw Data ----
cat("Loading raw dataset...\n")
data <- read.csv("data/raw/data.csv")

# Ensure required proxy variables are present
# We will use 'AssignmentCompletion' as a proxy for "Practice Self-assessment"
# We will use 'ExamScore' and 'FinalGrade' as proxies for "Stronger learning skills"
# Let's subset data for relevant columns to keep things focused
data_clean <- data %>%
  select(Age, Gender, StudyHours, Attendance, AssignmentCompletion, ExamScore, FinalGrade)

# ---- 2. Handle Missing Values ----
cat("Handling missing values...\n")
initial_rows <- nrow(data_clean)
data_clean <- data_clean %>% drop_na()
cat(sprintf("Removed %d rows with missing values.\n", initial_rows - nrow(data_clean)))

# ---- 3. Create Custom Variables (Feature Engineering) ----
cat("Creating derived variables for Self-Assessment...\n")
# Categorise Self-Assessment practice based on Assignment Completion rates
data_clean <- data_clean %>%
  mutate(
    SelfAssessment_Level = case_when(
      AssignmentCompletion >= 80 ~ "High",
      AssignmentCompletion >= 50 ~ "Moderate",
      TRUE ~ "Low"
    ),
    # Convert to factor with ordered levels
    SelfAssessment_Level = factor(SelfAssessment_Level, levels = c("Low", "Moderate", "High"))
  )

# ---- 4. Remove Outliers (IQR Method for ExamScore) ----
cat("Detecting and treating outliers in ExamScore...\n")
Q1 <- quantile(data_clean$ExamScore, 0.25)
Q3 <- quantile(data_clean$ExamScore, 0.75)
IQR <- Q3 - Q1
lower_bound <- Q1 - 1.5 * IQR
upper_bound <- Q3 + 1.5 * IQR

data_outliers_removed <- data_clean %>%
  filter(ExamScore >= lower_bound & ExamScore <= upper_bound)
cat(sprintf("Removed %d outliers from ExamScore.\n", nrow(data_clean) - nrow(data_outliers_removed)))

# ---- 5. Save Processed Data ----
cat("Saving cleaned dataset...\n")
dir.create("data/processed", showWarnings = FALSE)
write.csv(data_outliers_removed, "data/processed/cleaned_data.csv", row.names = FALSE)

cat("============================================================\n")
cat("Data cleaning complete. Proceed to 02_descriptive.R\n")
cat("============================================================\n")

File: scripts/01_data_cleaning.py
import pandas as pd
import numpy as np
import os

print("============================================================")
print("IT3011 Group Assignment - Script 01: Data Cleaning (Python)")
print("Learning Outcomes: LO1, LO4")
print("============================================================\n")

# ---- 1. Load Raw Data ----
print("Loading raw dataset...")
try:
    data = pd.read_csv("data/raw/data.csv")
except FileNotFoundError:
    print("Error: data/raw/data.csv not found!")
    exit(1)

# Subset required columns
columns_to_keep = ['Age', 'Gender', 'StudyHours', 'Attendance', 'AssignmentCompletion', 'ExamScore', 'FinalGrade']
data_clean = data[columns_to_keep].copy()

# ---- 2. Handle Missing Values ----
print("Handling missing values...")
initial_rows = len(data_clean)
data_clean = data_clean.dropna()
print(f"Removed {initial_rows - len(data_clean)} rows with missing values.")

# ---- 3. Create Custom Variables (Feature Engineering) ----
print("Creating derived variables for Self-Assessment...")

def categorize_completion(val):
    if val >= 80:
        return "High"
    elif val >= 50:
        return "Moderate"
    else:
        return "Low"

data_clean['SelfAssessment_Level'] = data_clean['AssignmentCompletion'].apply(categorize_completion)
data_clean['SelfAssessment_Level'] = pd.Categorical(data_clean['SelfAssessment_Level'], categories=["Low", "Moderate", "High"], ordered=True)

# ---- 4. Remove Outliers (IQR Method for ExamScore) ----
print("Detecting and treating outliers in ExamScore...")
Q1 = data_clean['ExamScore'].quantile(0.25)
Q3 = data_clean['ExamScore'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

data_outliers_removed = data_clean[(data_clean['ExamScore'] >= lower_bound) & (data_clean['ExamScore'] <= upper_bound)]
print(f"Removed {len(data_clean) - len(data_outliers_removed)} outliers from ExamScore.")

# ---- 5. Save Processed Data ----
print("Saving cleaned dataset...")
os.makedirs("data/processed", exist_ok=True)
data_outliers_removed.to_csv("data/processed/cleaned_data.csv", index=False)

print("\nData cleaning complete. Proceed to 02_descriptive.py")
